modules/custom/az_core/src/Plugin/ConfigProvider/QuickstartConfigProvider.php

```diff
@@ -27,7 +27,133 @@ class QuickstartConfigProvider extends ConfigProviderBase {
+          // Remove permissions that don't exist.
+          $profile_perms = array_intersect($profile_data['permissions'], $new_perms);
+          // Remove permissions that already are used.
+          $profile_perms = array_diff($profile_perms, $current_perms);
+          sort($profile_perms);
+
+          // Message about permissions.
+          foreach ($profile_perms as $perm) {
+            // @todo Use injection on user.permissions.
+            // @phpstan-ignore-next-line
+            \Drupal::messenger()->addMessage(t("Added permission %perm to %label",
+            [
+              '%perm' => $perm,
+              '%label' => $label,
+            ]));
+          }
+          if (!empty($profile_perms)) {
+            $role_config[$key]['permissions'] = array_unique(array_merge($current_perms, $profile_perms));
+          }
+        }
+      }
+    }
+
+    return $this->trimPermissions($role_config);
+  }
+
+  /**
+   * Trim invalid permissions from configuration data.
+   *
+   * @param array $config
+   *   A list of the configuration data keyed by configuration object name.
+   *
+   * @return array
+   *   A list of the configuration data keyed by configuration object name.
+   */
+  protected function trimPermissions(array $config) {
+    // Get permissions defined.
+    // @todo Use injection on user.permissions.
+    // @phpstan-ignore-next-line
+    $permission_definitions = \Drupal::service('user.permissions')->getPermissions();
+    $permissions = array_keys($permission_definitions);
+
+    // Add the configuration changes.
+    foreach ($config as $name => &$value) {
+      // Is this a permission configuration file?
+      if (strpos($name, 'user.role.') === 0) {
+        // Dependencies TBD based on permissions.
+        $value['dependencies'] = [];
+        // Trim active permissions list to what's expected.
+        if (!empty($value['permissions'])) {
+          $value['permissions'] = array_intersect($permissions, $value['permissions']);
+          sort($value['permissions']);
+        }
+
+        // Add dependencies based on permissions.
+        foreach ($value['permissions'] as $perm) {
+          $value['dependencies']['module'][] = $permission_definitions[$perm]['provider'];
+          if (!empty($permission_definitions[$perm]['dependencies'])) {
+            foreach ($permission_definitions[$perm]['dependencies'] as $dependency_type => $list) {
+              foreach ($list as $dependency) {
+                $value['dependencies'][$dependency_type][] = $dependency;
+              }
+            }
+          }
+        }
+        // Make sure dependencies are unique.
+        foreach ($value['dependencies'] as $dependency_type => &$dependencies) {
+          $dependencies = array_values(array_unique($dependencies));
+          sort($dependencies);
+        }
+      }
+
+      // Add transformed config hash.
+      $value = $this->addDefaultConfigHash($value);
+    }
+
+    return $config;
```

Comment:
Potential performance improvement:

Statically cache the result of `\Drupal::service('user.permissions')->getPermissions()` since it seems likely that this service method is called more than once per request (also called on line 48) and, as far as I can tell, the service isn't caching the results at all.

You'd obviously would want the new variable to be added at the top of the file instead of right here but this was the best I could do easily via commit suggestions.

```suggestion
    return $config;
  }
  /**
   * List of permissions defined.
   *
   * @var array
   */
  protected $permissionDefinitions = [];

  /**
   * Get permissions defined.
   *
   * @return array
   *   An array of permission definitions.
   */
  protected function getPermissionDefinitions() {
    if (empty($this->permissionDefinitions)) {
      // @todo Use injection on user.permissions.
      // @phpstan-ignore-next-line
      $this->permissionDefinitions = \Drupal::service('user.permissions')->getPermissions();
    }
    return $this->permissionDefinitions;
  }
```

Reply:
We could alternatively consider adding the Drupal core patch from the referenced [Drupal.org issue](https://www.drupal.org/project/drupal/issues/2339487), but that's probably ultimately riskier / has the potential to affect a lot more things:
https://www.drupal.org/project/drupal/issues/2339487#comment-13663937